File: bench/perl/src/thousand_sep.pl
#!/usr/bin/env perl
use 5.012;
use warnings;
use Time::HiRes qw/time/;

# Idea: Split number into a pre-decimal separator part, and one where
# every 3rd digit needs to be followed by a ",".
# E.g.: 12345678901234 -> 12 and 345678901234

my $pre_compile = time;
my $splitRx   = qr/^(\d{1,3})((?:\d{3})*)\n$/;
my $replaceRx = qr/(\d{3})/;

# Start timing
my $start = time;

while (my $line = <>) {
    my ($first, $rest) = $line =~ $splitRx;

    print $first;
    if ($rest ne '') {
        $rest =~ s/$replaceRx/,$1/g;
        print $rest;
    }
    print "\n";
}

# End timing
my $end = time;
my $elaps = int(($end - $start) * 1000);
my $elaps_compile = int(($start - $pre_compile) * 1000);

print STDERR "\ncompilation (ms): ${elaps_compile}\n";
print STDERR "matching (ms):    ${elaps}\n"

File: bench/python/src/thousand_sep.py
#!/usr/bin/env python
import sys
import datetime
import re

pre_compile = datetime.datetime.now()

splitRx   = re.compile(r'^(\d{1,3})((?:\d{3})*)\n$')
replaceRx = re.compile(r'(\d{3})')

# Start timing
start = datetime.datetime.now()

for line in sys.stdin:
    m = splitRx.match(line)
    first = m.group(1)
    rest = m.group(2)

    sys.stdout.write(first)
    if rest:
        sys.stdout.write( replaceRx.sub(r',\1', rest) )
    sys.stdout.write("\n")

# End timing
end = datetime.datetime.now()

# Elapsed time
elaps = end - start
elaps_compile = start - pre_compile
elaps_ms = elaps.seconds * 1000 + elaps.microseconds / 1000
elaps_compile_ms = elaps_compile.seconds * 1000 + elaps_compile.microseconds / 1000

sys.stderr.write("\ncompilation (ms): %s\n" % str(elaps_compile_ms))
sys.stderr.write("matching (ms):    %s\n" % str(elaps_ms))
